source/extensions/filters/http/json_to_metadata/filter.cc

```diff
@@ -0,0 +1,391 @@
+                         Stats::Counter& non_json) {
+  if (!body) {
+    handleAllOnMissing(rules, is_request_reported_);
+    no_body.inc();
+    return;
+  }
+
+  absl::StatusOr<Json::ObjectSharedPtr> result =
+      Json::Factory::loadFromStringNoThrow(body->toString());
+  if (!result.ok()) {
+    ENVOY_LOG(debug, result.status().message());
+    non_json.inc();
+    handleAllOnError(rules, reported_flag);
+    return;
+  }
+
+  Json::ObjectSharedPtr body_json = std::move(result.value());
+  StructMap struct_map;
+  for (const auto& rule : rules) {
+    const auto& keys = rule.keys_;
+    Json::ObjectSharedPtr node = body_json;
+    bool on_missing = false;
+    for (unsigned long i = 0; i < keys.size() - 1; i++) {
+      absl::StatusOr<Json::ObjectSharedPtr> next_node_result = node->getObjectNoThrow(keys[i]);
+      if (!next_node_result.ok()) {
+        ENVOY_LOG(warn, result.status().message());
+        handleOnMissing(rule, struct_map);
+        on_missing = true;
+        break;
+      }
+      node = std::move(next_node_result.value());
+    }
+    if (on_missing) {
+      continue;
+    }
+    absl::Status result = handleOnPresent(std::move(node), keys.back(), rule, struct_map);
+    if (!result.ok()) {
+      ENVOY_LOG(warn, fmt::format("{} key: {}", result.message(), keys.back()));
+      handleOnMissing(rule, struct_map);
+    }
+  }
+  success.inc();
+
+  finalizeDynamicMetadata(*decoder_callbacks_, struct_map, reported_flag);
+}
+
+void Filter::processRequestBody() {
+  processBody(decoder_callbacks_->decodingBuffer(), config_->requestRules(), is_request_reported_,
+              config_->stats().rq_success_, config_->stats().rq_no_body_,
+              config_->stats().rq_invalid_json_body_);
+}
+
+Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers, bool end_stream) {
+  ASSERT(config_->doRequest());
+  if (!config_->requestContentTypeAllowed(headers.getContentTypeValue())) {
+    is_request_reported_ = true;
+    config_->stats().rq_mismatched_content_type_.inc();
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  if (end_stream) {
+    handleAllOnMissing(config_->requestRules(), is_request_reported_);
+    config_->stats().rq_no_body_.inc();
+    return Http::FilterHeadersStatus::Continue;
+  }
+  return Http::FilterHeadersStatus::StopIteration;
+}
+
+Http::FilterDataStatus Filter::decodeData(Buffer::Instance&, bool end_stream) {
+  ASSERT(config_->doRequest());
+  if (is_request_reported_) {
+    return Http::FilterDataStatus::Continue;
+  }
+
+  if (end_stream) {
+    if (!decoder_callbacks_->decodingBuffer() ||
+        decoder_callbacks_->decodingBuffer()->length() == 0) {
+      handleAllOnMissing(config_->requestRules(), is_request_reported_);
+      config_->stats().rq_no_body_.inc();
+      return Http::FilterDataStatus::Continue;
+    }
+    processRequestBody();
+    return Http::FilterDataStatus::Continue;
+  }
+
+  if (decoder_callbacks_->decodingBuffer() &&
```

Comment:
We should also take into consideration the decoding buffer passed to the function, it's possible that that buffer != `decodingBuffer()` and we can miss triggering this if condition (especially if we leverage setting the decoding buffer limit.

For example, if 100B is our limit:
```
decodeData(80 byte input, no end stream) ; decodingBuffer() size: 0
decodeData(40 byte input, no end stream); decodingBuffer() size: 80
<-- we won't get another call of decodeData as when we tell it to buffer the 40, we should have 120 now in decodingBuffer 
```

Reply:
I'm not 100% sure but while we get `end_stream == true` the `decodingBuffer()` will be the full buffer? I was guessing the `data` of the parameter `Filter::decodeData(Buffer::Instance& data` is a *buffer_view* on `decodingBuffer()` . It would be hard to manage the status if `buffer != decodingBuffer()` I'm open to suggestion on how to make it right.

On the other hand, should we suggest filter users to set `per_connection_buffer_limit_bytes` and remove the `bufferLimitBytes` support in the filter, since the buffer limit is to avoid oom and connection level limitation is probably better. Although one big request could punish other light-weight request, we have overload manager to mitigate that.

wdyt @KBaichoo ?

Reply:
remove for now